source/slang/legalize-types.h

```diff
@@ -328,20 +328,139 @@ RefPtr<TypeLayout> getDerefTypeLayout(
 
 RefPtr<VarLayout> getFieldLayout(
     TypeLayout*     typeLayout,
-    String const&   mangledFieldName);
+    IRInst*         fieldKey);
 
-// Represents the "chain" of declarations that
-// were followed to get to a variable that we
-// are now declaring as a leaf variable.
+    /// Represents a "chain" of variables leading to some leaf field.
+    ///
+    /// Consider code like:
+    ///
+    ///     struct Branch { int leaf; }
+    ///     struct Tree { Branch left; Branch right; }
+    ///     cbuffer Forest
+    ///     {
+    ///         int maxTreeHeight;
+    ///         Tree tree;
+    ///     }
+    ///
+    /// If we ask "what is the offset of `leaf`" the simple answer is zero,
+    /// but sometimes we are talking about `Forest.tree.right.leaf` which
+    /// will have a very different offset. In Slang parameters can consume
+    /// various (and multiple) resource kinds, so a single offset can't
+    /// be tunneled down through most recursive procedures.
+    ///
+    /// Instead we use a "chain" that works up through the stack, and
+    /// records the path from leaf field like `leaf` up to whatever
+    /// variable is the root for the curent operation.
+    ///
+    /// Operations like computing an offset can then be encoded by
+    /// starting with zero and then walking up the chain and adding in
+    /// offsets as encountered.
+    ///
+struct SimpleLegalVarChain
+{
+    // The next link up the chain, or null if this is the end.
+    SimpleLegalVarChain*    next = nullptr;
+
+    // The layout for the variable at this link in thain.
+    VarLayout*              varLayout = nullptr;
+};
+
+    /// A "chain" of variable declarations that can handle both primary and "pending" data.
+    ///
+    /// In the presence of interface-type fields, a single variable may
+    /// have data that sits in two distinct allocations, and may have
+    /// `VarLayout`s that represent offseting into each of those
+    /// allocations.
+    ///
+    /// A `LegalVarChain` tracks two distinct `SimpleVarChain`s: one for
+    /// the primary/ordinary data allocation, and one for any pending
+    /// data.
+    ///
+    /// It is okay if the primary/pending chains have different numbers
+    /// of links in them.
+    ///
+    /// Offsets for particular resource kinds in the primary or pending
+    /// data allocation can be queried on the appropriate sub-chain.
+    ///
 struct LegalVarChain
 {
-    LegalVarChain*  next;
-    VarLayout*      varLayout;
+    // The chain of variables that represents the primary allocation.
+    SimpleLegalVarChain*    primaryChain = nullptr;
```

Comment:
Can of worms indeed. :)

I'd argue that a "bag of bits" type can still have constructors; they are just convenience methods for initializing the bag of bits.

Also, in this case inheritance isn't being used for OOP reasons, but just as a convenient way to streamline the amount of code that has to be written by having an RAII-style type for constructing a suitable value on the stack, that can then be passed along as the type it constructs. We could replace the inheritance with a field + an explicit accessor, or an implicit conversion function, but it just seems like extra complexity to me.

I take the point that the type is a bit hard to explain when it appears in a public header, since it is really an implementation detail of very specific algorithms.